phoenix-core/src/test/java/org/apache/phoenix/query/KeyRangeMoreTest.java

```diff
@@ -192,64 +193,78 @@ public void testListIntersectForBoundary() throws Exception {
+                Arrays.asList(3, 5),
+                Arrays.asList(true, true));
         listIntersectAndAssert(rowKeyRanges1, rowKeyRanges2, expected);
     }
 
+    @Test
+    public void testKeyRangeCompareUpperRange() throws Exception {
+        List<KeyRange> rowKeyRanges1 = CreateKeyRangeListWithFixedLowerRange(
+                Arrays.asList(Integer.MAX_VALUE, Integer.MAX_VALUE, 10000, 1001, 1000, 1000, 1000, 1000),
+                Arrays.asList(false, false, true, true, true, true, false, true));
+        List<KeyRange> rowKeyRanges2 = CreateKeyRangeListWithFixedLowerRange(
+                Arrays.asList(Integer.MAX_VALUE, 10000, Integer.MAX_VALUE, 1000, 1001, 1000, 1000, 1000),
+                Arrays.asList(false, false, false, true, true, true, false, false));
+        List<Integer> expectedResults = Arrays.asList(0, 1, -1, 1, -1, 0, 0, 1);
+        assertEquals(rowKeyRanges1.size(), rowKeyRanges2.size());
+        assertEquals(rowKeyRanges1.size(), expectedResults.size());
+
+        for (int i = 0; i < expectedResults.size(); i++) {
+            int compareResult = KeyRange.compareUpperRange(rowKeyRanges1.get(i), rowKeyRanges2.get(i));
+            assertEquals(expectedResults.get(i).intValue(), compareResult);
+        }
+    }
+
+    private static List<KeyRange> CreateKeyRangeListWithFixedLowerRange(List<Integer> keys, List<Boolean> boundaryConditions) {
+        assertEquals(keys.size(), boundaryConditions.size());
+        List<Integer> newKeys = Lists.newArrayListWithCapacity(keys.size() * 2);
+        List<Boolean> newBoundaryConditions = Lists.newArrayListWithCapacity(boundaryConditions.size() * 2);
+
+        for (int i = 0; i < keys.size(); i++) {
+            newKeys.add(0);
+            newBoundaryConditions.add(true);
+            newKeys.add(keys.get(i));
+            newBoundaryConditions.add(boundaryConditions.get(i));
+        }
+
+        return CreateKeyRangeList(newKeys, newBoundaryConditions);
+    }
+
+    private static List<KeyRange> CreateKeyRangeList(List<Integer> keys, List<Boolean> boundaryConditions) {
+        assertEquals(keys.size(), boundaryConditions.size());
+        assertTrue(keys.size() % 2 == 0);
+
+        int size = keys.size() / 2;
+        List<KeyRange> keyRangeList = Lists.newArrayListWithCapacity(size);
+
+        for (int i = 0; i < size; i++) {
+            byte[] startKey = keys.get(2*i).equals(Integer.MIN_VALUE) ? KeyRange.UNBOUND : PInteger.INSTANCE.toBytes(keys.get(2*i));
+            byte[] endKey = keys.get(2*i + 1).equals(Integer.MAX_VALUE) ? KeyRange.UNBOUND : PInteger.INSTANCE.toBytes(keys.get(2*i + 1));
+            keyRangeList.add(PInteger.INSTANCE.getKeyRange(startKey, boundaryConditions.get(2*i), endKey, boundaryConditions.get(2*i+1)));
+        }
+
+        return keyRangeList;
+    }
+
     private static void listIntersectAndAssert(List<KeyRange> rowKeyRanges1,List<KeyRange> rowKeyRanges2,List<KeyRange> expected) {
         for(int i=0;i<200;i++) {
```

Comment:
I see you didn't change this test case but as case as it is non-deterministic and the number of permutations are O(n!) which grows large even on small subsets beyond the scope of a normal rng period.

Reply:
The purpose of the original test author just want to randomly shuffle the ranges in list for certain times and make sure the function can correct sort the given tow ranges and calculate the intersection. I don't like this test code because once there is failure we don't know what's the case which caused the failure, but I think improving this is out of scope of my change due to the following two reasons:
1. Using the ranges in ascending order is efficient for testing my fix to the original problem.
2. Reshuffle the ranges should be the unit test of KeyRange.coalesce(...) instead of KeyRange.intersect().

Reply:
This is okay with me.